sdks/python/apache_beam/transforms/util_test.py

```diff
@@ -423,6 +425,70 @@ def test_reshuffle_streaming_global_window(self):
                 label='after reshuffle')
     pipeline.run()
 
+  @attr('ValidatesRunner')
+  def test_reshuffle_preserves_timestamps(self):
+    pipeline = TestPipeline()
+
+    # Create a PCollection and assign each element with a different timestamp.
+    before_reshuffle = (pipeline
+                        | "Four elements" >> beam.Create([
+                            {'name': 'foo', 'timestamp': MIN_TIMESTAMP},
+                            {'name': 'foo', 'timestamp': 0},
+                            {'name': 'bar', 'timestamp': 33},
+                            {'name': 'bar', 'timestamp': MAX_TIMESTAMP},
+                        ])
+                        | "With timestamp" >> beam.Map(
+                            lambda element: beam.window.TimestampedValue(
+                                element, element['timestamp'])))
+
+    # For each element in a PCollection, gets the current timestamp of the
+    # element and reassigns the timestamp to the element.
+    class AddTimestamp(beam.DoFn):
+      def process(self, element, timestamp=beam.DoFn.TimestampParam):
+        yield beam.window.TimestampedValue(element, timestamp)
+
+    # Reshuffle the PCollection above and assign the timestamp of an element to
+    # that element again.
+    after_reshuffle = (before_reshuffle
+                       | "Reshuffle" >> beam.Reshuffle()
+                       | "With timestamps again" >> beam.ParDo(AddTimestamp()))
+
+    # Given an element, emits a string which contains the timestamp and the name
+    # field of the element.
+    class FormatWithTimestamp(beam.DoFn):
```

Comment:
There's no need to create a DoFn, just create a function and do beam.Map(format_with_timestamp).

Reply:
I need beam.DoFn.TimestampParam to get the timestamp of each element.

Reply:
You can have a method

```
def format_with_timestamp(element, timestamp=beam.DoFn.TimestampParam):
    ...
```

rather than a full DoFn.

Reply:
Done. Thank you! It's good to know the code can be written in this way.